sys/analog_util/adc_util.c

```diff
@@ -14,28 +15,51 @@
  * @brief       ADC utility function implementation
  *
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Joakim Nohlgård <joakim.nohlgard@eistec.se>
  *
  * @}
  */
 
+#include <stdint.h>
+#include "cpu.h"
+#include "periph/adc.h"
 #include "analog_util.h"
 
-/* keep a max value to ADC resolution mapping for quick access in the ROM */
-static const int val_max[] = {
-    [ADC_RES_6BIT] =    0x003f,
-    [ADC_RES_8BIT] =    0x00ff,
-    [ADC_RES_10BIT] =   0x03ff,
-    [ADC_RES_12BIT] =   0x0fff,
-    [ADC_RES_14BIT] =   0x3fff,
-    [ADC_RES_16BIT] =   0xffff
-};
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+/**
+ * @brief Convert adc_res_t resolution setting into numeric bit count
+ */
+static unsigned int _adc_res_bits(adc_res_t res)
+{
+    switch (res) {
+        case ADC_RES_6BIT:
+            return 6;
+        case ADC_RES_8BIT:
+            return 8;
+        case ADC_RES_10BIT:
+            return 10;
+        case ADC_RES_12BIT:
+            return 12;
+        case ADC_RES_14BIT:
+            return 14;
+        case ADC_RES_16BIT:
+        default:
+            return 16;
+    }
+}
 
 int adc_util_map(int sample, adc_res_t res, int min, int max)
 {
-    return ((((max - min) * sample) / val_max[res]) + min);
+    /* Using 64 bit signed int as intermediate to prevent overflow when range
+     * multiplied by sample requires more bits than int offers */
+    int scaled = (((int64_t)(max - min) * sample) >> _adc_res_bits(res));
+    DEBUG("scaled: %d\n", scaled);
+    return (min + scaled);
 }
 
 float adc_util_mapf(int sample, adc_res_t res, float min, float max)
 {
-    return ((((max - min) * sample) / val_max[res]) + min);
+    return ((((max - min) * sample) / (1 << _adc_res_bits(res))) + min);
```

Comment:
As `_adc_res_bits(res)` can have a value of 16, I think this `1` should be set to be 32 bit wide or it will break on 16bit platforms.

Also right now, there is a by one difference from the previous version when dividing. Not sure if it matters but just noticing.

Reply:
Good catch on the integer width, will fix